diff --git a/libsolidity/codegen/ABIFunctions.cpp b/libsolidity/codegen/ABIFunctions.cpp
index 474848760..91816353e 100644
--- a/libsolidity/codegen/ABIFunctions.cpp
+++ b/libsolidity/codegen/ABIFunctions.cpp
@@ -1120,7 +1120,7 @@ std::string ABIFunctions::abiDecodingFunctionValueType(Type const& _type, bool _
 		Whiskers templ(R"(
 			function <functionName>(offset, end) -> value {
 				value := <load>(offset)
-				<validator>(value)
+//				<validator>(value)
 			}
 		)");
 		templ("functionName", functionName);
diff --git a/scripts/build.sh b/scripts/build.sh
index 6edd60bd9..4189279b8 100755
--- a/scripts/build.sh
+++ b/scripts/build.sh
@@ -20,7 +20,7 @@ cd "${BUILDDIR}"
 cmake .. -DCMAKE_BUILD_TYPE="$BUILD_TYPE" "${@:2}"
 make -j2
 
-if [[ "${CI}" == "" ]]; then
-	echo "Installing ..."
-	sudo make install
-fi
+#if [[ "${CI}" == "" ]]; then
+#	echo "Installing ..."#
+#	sudo make install
+#fi
